// stryker/src/reporters/ProgressKeeper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var report_1 = require("stryker-api/report");
var ProgressKeeper = /** @class */ (function () {
    function ProgressKeeper() {
        this.progress = {
            survived: 0,
            tested: 0,
            total: 0
        };
    }
    ProgressKeeper.prototype.onAllMutantsMatchedWithTests = function (matchedMutants) {
        this.progress.total = matchedMutants.filter(function (m) { return m.scopedTestIds.length > 0; }).length;
    };
    ProgressKeeper.prototype.onMutantTested = function (result) {
        this.progress.tested++;
        switch (result.status) {
            case report_1.MutantStatus.NoCoverage:
                this.progress.tested--; // correct for not tested, because no coverage
                break;
            case report_1.MutantStatus.Survived:
                this.progress.survived++;
                break;
        }
    };
    return ProgressKeeper;
}());
exports.default = ProgressKeeper;
